pelican/utils.py

```diff
@@ -622,79 +629,117 @@ def process_translations(content_list, order_by=None):
-            logger.warning('All items with slug "%s" are translations', slug)
-            candidate_items = items
-
-        # find items with default language
-        original_items = list(filter(
-            attrgetter('in_default_lang'),
-            candidate_items))
-
-        # if there is no article with default language, go back one step
-        if not original_items:
-            original_items = candidate_items
-
-        # display warning if there are several original items
-        if len(original_items) > 1:
-            _warn_source_paths(
-                'There are %s original (not translated) items with slug "%s"',
-                original_items,
-                slug)
-
-        index.extend(original_items)
-        translations.extend([x for x in items if x not in original_items])
-        for a in items:
-            a.translations = [x for x in items if x != a]
+            warnings.append(_get_warning_with_source_paths(
+                'There are {} items with empty slug',
+                items))
+
+        original_items, slug_warnings = get_original_items(items, slug)
+        # if there are several original items, use secondary_attr to
+        # differentiate if possible
+        if len(original_items) > 1 and secondary_attr:
+            items.sort(key=attrgetter(secondary_attr))
+            grouped_secondarily = groupby(items, attrgetter(secondary_attr))
+            for val, val_items in grouped_secondarily:
+                val_items = list(val_items)
+                with_slug_and_val_str = '{} and {} "{}"'.format(
+                    with_slug_str, secondary_attr, val)
+                original_cat_items, slug_cat_warnings = get_original_items(
+                    val_items, with_slug_and_val_str)
+                warnings.extend(slug_cat_warnings)
+                index.extend(original_cat_items)
+
+                for a in val_items:
+                    a.translations = [x for x in val_items if x != a]
+        else:
+            index.extend(original_items)
+            warnings.extend(slug_warnings)
+            for a in items:
+                a.translations = [x for x in items if x != a]
+
+    translations = [x for x in content_list if x not in index]
+    for warning in warnings:
+        logger.warning(warning)
+
+    return index, translations
+
+
+def get_original_items(items, with_str):
+    warnings = []
+    # warn if several items have the same lang
+    for lang, lang_items in groupby(items, attrgetter('lang')):
+        lang_items = list(lang_items)
+        if len(lang_items) > 1:
+            warnings.append(_get_warning_with_source_paths(
+                'There are {} items {} with lang {}',
+                lang_items,
+                with_str,
+                lang))
+
+    # items with `translation` metadata will be used as translations...
+    candidate_items = list(filter(
+        lambda i:
+        i.metadata.get('translation', 'false').lower() == 'false',
+        items))
+
```

Comment:
Since you're rewriting stuff, can we replace `list(filter(...))` with proper list comprehension?

Reply:
Sure